src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/AbstractEntityAIGuard.java

```diff
@@ -590,6 +595,52 @@ public IAIState patrol()
         return GUARD_PATROL;
     }
 
+    /**
+     * Patrol between all completed nodes in the assigned mine
+     * @return the next point to patrol to
+     */
+    public IAIState patrolMine()
+    {
+        if (buildingGuards.getMinePos() == null)
+        {
+            return PREPARING;
+        }
+        if (currentPatrolPoint == null ||  worker.isWorkerAtSiteWithMove(currentPatrolPoint, 2))
+        {
+            final IBuilding building = buildingGuards.getColony().getBuildingManager().getBuilding(buildingGuards.getMinePos());
+            if (building != null && building instanceof BuildingMiner)
+            {
+                final BuildingMiner miner = (BuildingMiner) building;
+                final Level level = miner.getCurrentLevel();
+                if (level == null || level.getNodes().isEmpty())
+                {
+                    setNextPatrolTarget(miner.getPosition());
+                }
+                else
+                {
+                    final List<Map.Entry<Vec2i, Node>> filteredNodes = level.getNodes().entrySet()
```

Comment:
I would add a method inside the level with sth like "getRandomCompletedNode" or so.
Potentially we could do a greedy calc here then.
Just get any random node, and then if that one is valid return it, and if not just go to its parent and check the same. This is O(1) in an average case and O(logn) in the worst case instead of the O(N)